pydantic_ai_slim/pydantic_ai/models/ollama.py

```diff
@@ -0,0 +1,126 @@
+from __future__ import annotations as _annotations
+
+from collections.abc import Mapping, Sequence
+from dataclasses import dataclass, field
+
+from httpx import AsyncClient as AsyncHTTPClient
+
+from . import (
+    AbstractToolDefinition,
+    AgentModel,
+    Model,
+    cached_async_http_client,
+    check_allow_model_requests,
+)
+
+try:
+    from openai import AsyncOpenAI
+    from openai.types import chat
+except ImportError as e:
+    raise ImportError(
+        'Please install `openai` to use the OpenAI model, '
+        "you can use the `openai` optional group — `pip install 'pydantic-ai[openai]'`"
+    ) from e
+
+# Only required in the below typing is enforced
+# try:
+#     from ollama import list, ListResponse
+# except ImportError as e:
+#     raise ImportError(
+#         'Please install `ollama` to use the Ollama model, '
+#         "you can use the `ollama` optional group — `pip install 'pydantic-ai[ollama]'`"
+#     ) from e
+
+
+from .openai import OpenAIAgentModel
+
+# NB: Currently commented out but this is
+# a possible approach to type the ollama model names
+# # This is really nasty, but ollama doesn't
+# # expose a type of all available models, and as
+# # there are 100's and being udpated all the time,
+# # it seems difficult to maintain a list of all possible
+# # versions statically.
+# # This code gets the ones that are _currently_ available
+# # to the user as they are ones that they have already downloaded
+# ollama_ls: ListResponse = list()
+# ollama_available_models: list[str] = [x.model for x in ollama_ls.models]
+# OllamaModelName = Literal[tuple(ollama_available_models)]
+
+
+# Inherits from this model
+@dataclass(init=False)
+class OllamaModel(Model):
+    """A model that implements Ollama using the OpenAI API.
+
+    Internally, this uses the [OpenAI Python client](https://github.com/openai/openai-python) to interact with the Ollama server.
+
+    Apart from `__init__`, all methods are private or match those of the base class.
+    """
+
+    model_name: str
+    client: AsyncOpenAI = field(repr=False)
```

Comment:
you can probably just make this

```py
openai_model: OpenAIModel
```

and thereby defer even more logic to use `OpenAIModel`.

Reply:
I'm a little confused by this comment 😅 

are you effectively saying get rid of the `model_name` and `client` and replace it with an `openai_model`, and then in the `__init__`, have something like this?

```py
from .openai import OpenAIAgentModel, OpenAIModel

@dataclass(init=False)
class OllamaModel(Model):

    openai_model: OpenAIModel

    def __init__(
        self,
        model_name: str,
        *,
        base_url: str = 'http://localhost:11434/v1/',
        openai_client: AsyncOpenAI | None = None,
        http_client: AsyncHTTPClient | None = None,
    ):
        self.openai_model = OpenAIModlel(model_name=model_name, openai_client=openai_client, http_client=http_client, api_key='ollama')
        if openai_client is not None:
            assert http_client is None, 'Cannot provide both `openai_client` and `http_client`'
            self.openai_model = OpenAIModlel(model_name=model_name, openai_client=openai_client, http_client=http_client)
        elif http_client is not None:
            # API key is not required for ollama but a value is required to create the client
            oai_client = AsyncOpenAI(base_url=base_url, api_key='ollama', http_client=http_client)
            self.openai_model = OpenAIModlel(model_name=model_name, openai_client=oai_client, http_client=http_client)
        else:
            # API key is not required for ollama but a value is required to create the client
            oai_client = AsyncOpenAI(base_url=base_url, api_key='ollama', http_client=cached_async_http_client())
            self.openai_model = OpenAIModlel(model_name=model_name, openai_client=oai_client, http_client=http_client)
        
    async def agent_model(
        self,
        function_tools: Mapping[str, AbstractToolDefinition],
        allow_text_result: bool,
        result_tools: Sequence[AbstractToolDefinition] | None,
    ) -> AgentModel:
        
        return self.openai_model.agent_model(
            function_tools=function_tools,
            allow_text_result=allow_text_result, 
            result_tools=result_tools,
        )
```

Reply:
basically yes, you don't need `_map_tool_definition`

Reply:
Ok great @samuelcolvin . 

I've udpated. I've also created a `OpenAIModelName` type which is a `Union[ChatModel,str]`, as that type is used quite a lot now due to the refactor